### PR TITLE
fix for issue 28

### LINKED ISSUE #28: WSDL not parseable
**Description:**
I want to transform a specific WSDL to OpenAPI by using this library. In general, it simply works, but in one case, the WSDL file is not parsed. SoapUI is reading the WSDL file without problems.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**
soap-to-rest v1.7

**OS, DB, other environment details and versions:**
Ubuntu 22.04, OracleJDK 17.0.12   

**Steps to reproduce:**
Try to convert this example WSDL:
<details>
  <summary>buggy WSDL</summary>
  
  ```xml
<wsdl:definitions xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/"
  xmlns:tns="urn:ServerWsdl" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/"
  xmlns:xsd="http://www.w3.org/2001/XMLSchema" name="myServer"
  targetNamespace="urn:ServerWsdl">
  <wsdl:types>
    <xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema" targetNamespace="urn:ServerWsdl">
      <xsd:complexType name="getRequestType">
        <xsd:sequence>
          <xsd:element name="header" type="tns:standardheaderType" />
          <xsd:element name="body" type="tns:getRequestBodyType" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="getResultType">
        <xsd:sequence>
          <xsd:element name="result" type="tns:resultType" maxOccurs="1" minOccurs="0" />
          <xsd:element name="error" type="tns:errorType" maxOccurs="1" minOccurs="0" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="standardheaderType">
        <xsd:sequence>
          <xsd:element name="applicationversionnr" type="xsd:string" maxOccurs="1" minOccurs="1" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="getRequestBodyType">
        <xsd:sequence>
          <xsd:element name="transactioncode" type="xsd:string" maxOccurs="1" minOccurs="1" />
          <xsd:element name="languageisocode" type="xsd:string" maxOccurs="1" minOccurs="1" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="disclosureinformationType">
        <xsd:sequence>
          <xsd:element name="companyname" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element
            name="street" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element
            name="countryisocode" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element
            name="city" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element name="postcode"
            type="xsd:string" maxOccurs="1" minOccurs="0" /> 
</xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="clientinformationType">
        <xsd:sequence>
          <xsd:element name="fieldname" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element name="fieldvalue" type="xsd:string" maxOccurs="1" minOccurs="0" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="resultType">
        <xsd:sequence>
          <xsd:element name="transactioncode" type="xsd:string" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="errorType">
        <xsd:sequence>
          <xsd:element name="errorcode" type="xsd:string" />
          <xsd:element name="errortext" type="xsd:string" />
        </xsd:sequence>
      </xsd:complexType>
    </xsd:schema>
  </wsdl:types>
  <wsdl:message name="getRequest">
    <wsdl:part name="getRequest" type="tns:getRequestType" />
  </wsdl:message>
  <wsdl:message name="getResponse">
    <wsdl:part name="getResponse" type="tns:getResultType" />
  </wsdl:message>
  <wsdl:portType name="myServer">
    <wsdl:operation name="get">
      <wsdl:input message="tns:getRequest" />
      <wsdl:output message="tns:getResponse" />
    </wsdl:operation>
  </wsdl:portType>
  <wsdl:binding name="myServerSOAP" type="tns:myServer">
    <soap:binding style="rpc" transport="http://schemas.xmlsoap.org/soap/http" />
    <wsdl:operation name="get">
      <soap:operation soapAction="urn:ServerWsdl/get" />
      <wsdl:input>
        <soap:body namespace="urn:ServerWsdl" use="literal" />
      </wsdl:input>
      <wsdl:output>
        <soap:body namespace="urn:ServerWsdl" use="literal" />
      </wsdl:output>
    </wsdl:operation>
  </wsdl:binding>
  <wsdl:service name="myServer">
    <wsdl:port binding="tns:myServerSOAP" name="myServerSOAP">
      <soap:address
        location="https://some.server.de/myServer.php" />
    </wsdl:port>
  </wsdl:service>
</wsdl:definitions>
  ```
</details>

I then get the following exception:
> java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 0
>	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4608)
>	at java.base/java.lang.String.substring(String.java:2711)
>	at > org.wso2.soaptorest.SOAPRequestBodyGenerator.createSOAPRequestXMLForOperation(SOAPRequestBodyGenerator.java:355)
>	at > org.wso2.soaptorest.SOAPRequestBodyGenerator.generateSOAPtoRESTConversionObjectFromOAS(SOAPRequestBodyGenerat> or.java:150)
>	at org.wso2.soaptorest.SOAPToRESTConverter.getSOAPtoRESTConversionData(SOAPToRESTConverter.java:51)

I took a deeper look at the buggy WSDL and I finally got a WSDL which is perfectly working:
<details>
  <summary>working WSDL</summary>
  
  ```xml
<wsdl:definitions xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/"
  xmlns:tns="urn:ServerWsdl" xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/"
  xmlns:xsd="http://www.w3.org/2001/XMLSchema" name="myServer"
  targetNamespace="urn:ServerWsdl">
  <wsdl:types>
    <xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema" targetNamespace="urn:ServerWsdl">
      <xsd:complexType name="getRequestType">
        <xsd:sequence>
          <xsd:element name="header" type="tns:standardheaderType" />
          <xsd:element name="body" type="tns:getRequestBodyType" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="getResultType">
        <xsd:sequence>
          <xsd:element name="result" type="tns:resultType" maxOccurs="1" minOccurs="0" />
          <xsd:element name="error" type="tns:errorType" maxOccurs="1" minOccurs="0" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="standardheaderType">
        <xsd:sequence>
          <xsd:element name="applicationversionnr" type="xsd:string" maxOccurs="1" minOccurs="1" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="getRequestBodyType">
        <xsd:sequence>
          <xsd:element name="transactioncode" type="xsd:string" maxOccurs="1" minOccurs="1" />
          <xsd:element name="languageisocode" type="xsd:string" maxOccurs="1" minOccurs="1" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="disclosureinformationType">
        <xsd:sequence>
          <xsd:element name="companyname" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element
            name="street" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element
            name="countryisocode" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element
            name="city" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element name="postcode"
            type="xsd:string" maxOccurs="1" minOccurs="0" />
</xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="clientinformationType">
        <xsd:sequence>
          <xsd:element name="fieldname" type="xsd:string" maxOccurs="1" minOccurs="0" />
          <xsd:element name="fieldvalue" type="xsd:string" maxOccurs="1" minOccurs="0" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="resultType">
        <xsd:sequence>
          <xsd:element name="transactioncode" type="xsd:string" />
        </xsd:sequence>
      </xsd:complexType>
      <xsd:complexType name="errorType">
        <xsd:sequence>
          <xsd:element name="errorcode" type="xsd:string" />
          <xsd:element name="errortext" type="xsd:string" />
        </xsd:sequence>
      </xsd:complexType>
    </xsd:schema>
  </wsdl:types>
  <wsdl:message name="getRequest">
    <wsdl:part name="getRequest" type="tns:getRequestType" />
  </wsdl:message>
  <wsdl:message name="getResponse">
    <wsdl:part name="getResponse" type="tns:getResultType" />
  </wsdl:message>
  <wsdl:portType name="myServer">
    <wsdl:operation name="get">
      <wsdl:input message="tns:getRequest" />
      <wsdl:output message="tns:getResponse" />
    </wsdl:operation>
  </wsdl:portType>
  <wsdl:binding name="myServerBinding" type="tns:myServerWsdl">
    <soap:binding style="rpc" transport="http://schemas.xmlsoap.org/soap/http" />
    <wsdl:operation name="get">
      <soap:operation soapAction="urn:ServerWsdl/get" />
      <wsdl:input>
        <soap:body namespace="urn:ServerWsdl" use="literal" />
      </wsdl:input>
      <wsdl:output>
        <soap:body namespace="urn:ServerWsdl" use="literal" />
      </wsdl:output>
    </wsdl:operation>
  </wsdl:binding>
  <wsdl:service name="myServer">
    <wsdl:port binding="tns:myServerBinding" name="myServerSOAP">
      <soap:address
        location="https://some.server.de/myServer.php" />
    </wsdl:port>
  </wsdl:service>
</wsdl:definitions>
  ```
</details>
The difference is, that I changed in wsdl:service -> wsdl:port binding to "tns:myServerBinding" and wsdl:binding is now 
```xml
<wsdl:binding name="myServerBinding" type="tns:myServerWsdl"> 
```
Because the original WSDL is parsed by SoapUI, it might be a bug in this library. Could you please have a look at it?

**Related Issues:**

### PR DESCRIPTION
**Pull request by indika-dev**
## Automation tests
 - Unit tests 
   > added org.wso2.soaptorest.SOAPToRESTConverterTest.testIssue28() and src/test/resources/issue-28/failing.wsdl from bugreport
 - Integration tests
   > N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: src/test/java/org/wso2/soaptorest/SOAPToRESTConverterTest.java [hunk 1]
```diff
@@ -1,57 +1,70 @@
 /*
  * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
  *
- * WSO2 Inc. licenses this file to you under the Apache License,
- * Version 2.0 (the "License"); you may not use this file except
- * in compliance with the License.
- * You may obtain a copy of the License at
+ * WSO2 Inc. licenses this file to you under the Apache License, Version 2.0 (the "License"); you
```
Comment (rosensilva) [inline src/test/java/org/wso2/soaptorest/SOAPToRESTConverterTest.java]: Would it be possible to send only the code changes? This pull request contains a lot of code formatting, which makes it difficult to discern the specific changes you made.

**Comment by rosensilva** [comment]
Hi @indika-dev,

Thank you for your valuable contribution. Could you please address the review comment?
Regards,
Rosen

**Comment by indika-dev** [comment]
Do you mean the conflict with .gitignore?

Kind regards,
Stefan

**Comment by rosensilva** [comment]
when we check out the [src/main/java/org/wso2/soaptorest/SOAPRequestBodyGenerator.java](https://github.com/wso2/soap-to-rest/pull/29/files#diff-31fe1e911631e34920269a6cce07783b01353b75cad93847884cc2933a8ece8d) file, we can see that there are 724 lines of code changes. It's a bit tricky to figure out the actual changes just by looking at the PR. This PR includes both code formatting git diff and real fix diff. Ideally, if we're fixing the formatting, it would be great to do it in a separate commit.

Thanks!

**Comment by indika-dev** [comment]
I'm very sorry for the inconvinience. I recommitted the changed files without reformatting the whole file. Please have a look at it.

**Comment by rosensilva** [comment]
Hi @indika-dev ,

Thanks for making the suggested change.

**Review by rosensilva — APPROVED** [review]

**Pull request merged**